File: src/dev/coralnpu/uart_console.cc
#include "dev/coralnpu/uart_console.hh"

#include <cstdio>
#include <cstring>

#include "mem/packet.hh"
#include "mem/packet_access.hh"

namespace gem5 {

UartConsole::UartConsole(const Params &p)
    : BasicPioDevice(p, p.pio_size)
{}

Tick
UartConsole::read(PacketPtr pkt)
{
    // TX-ready / no RX FIFO — return zeros for any read size.
    // memset handles packets of any size (1-byte status reads, 32-byte
    // L1I cache-line fills from speculative instruction fetches, etc.).
    memset(pkt->getPtr<uint8_t>(), 0, pkt->getSize());
    pkt->makeAtomicResponse();
    return pioDelay;
}

Tick
UartConsole::write(PacketPtr pkt)
{
    if (pkt->getSize() >= 1) {
        char c = static_cast<char>(pkt->getLE<uint8_t>());
        putchar(c);
        if (c == '\n')
            fflush(stdout);
    }
    pkt->makeAtomicResponse();
    return pioDelay;
}

} // namespace gem5

File: src/dev/coralnpu/UartConsole.py
from m5.objects.Device import BasicPioDevice
from m5.params import Param


class UartConsole(BasicPioDevice):
    """
    Minimal MMIO UART-TX console for CoralNPU gem5 simulation.

    Maps a small address window (default: 0xFFFFFFF8, size 8 B).
    A byte write to any address in the window is printed to host stdout,
    emulating the firmware's UART character-output idiom:
        sb  a0, 0xFFFFFFF8
    """

    type = "UartConsole"
    cxx_header = "dev/coralnpu/uart_console.hh"
    cxx_class = "gem5::UartConsole"

    pio_size = Param.Addr(8, "Size of the MMIO window in bytes")
